### ISSUE TITLE
Ability to filter by "user-perceived" unhandled exceptions/fatal errors generated by the Kotlin SDK

### ISSUE THREAD

**Issue by arifken** [getsentry/sentry-java#3425]
### Problem Statement

Due to the behavior of [coroutines exception handling](https://kotlinlang.org/docs/exception-handling.html), a failure to catch the exception can sometimes result in a crash, if the request was made from certain contexts, and can be swallowed when made from others. 

In both of these cases, Sentry labels the error as `level:fatal` with `exception.handled:false`. However, this creates the possibility for a spike in "crashes" that do not actually crash the app for the user. 

We recently encountered this issue with a `android_getaddrinfo` exception, where a user was trying to make a network connection while in airplane mode. What we saw is a huge spike in errors in Sentry, but when we went to Google Play, we were able to _filter by user-perceived errors_ and see that the actual count of crashes for this issue was very very low. 

It appears that Sentry's "Crash free user rate" takes into consideration fatal errors that are "user perceived" vs. not (the spike in errors did not impact our crash free user rate).

### Solution Brainstorm

We'd love to be able to query by `user_perceived:true` in Discover, Dashboards, and Alerts (for error data sources) so that we can differentiate between user-perceived vs non-main thread crashes ourselves. 

when specifying `user_perceived:true`, I would expect to only see crashes that resulted in the Android app closing, and contribute to the crash-free user rate shown by Sentry.

when specifying `user_perceived:false level:fatal`, I would expect to only see errors that are swallowed up by the parent coroutine and do not actually crash the app for the user. 

### Product Area

Issues

**Comment by kahest** [comment]
@anthonycr @arifken we report fatally crashed sessions as `crashed`, and sessions that experience (potentially multiple) non-fatal errors as `errored`. You should be able to see the "Error free session rate" by looking at the [Release Details](https://docs.sentry.io/product/releases/release-details/). Crashing coroutines should be counted towards errored sessions.

Also, can you confirm if this is about the [Kotlin Multiplatform](https://github.com/getsentry/sentry-kotlin-multiplatform) SDK?

**Comment by romtsn** [comment]
> However, the app itself does not terminate, so the Sentry session does not close, but remains open. 

I'm not sure if this is correct, because we actually finish the session if our exception handler encountered a crash, see here: https://github.com/getsentry/sentry-java/blob/84892a8db59d768e89a04a05439f5198c162301b/sentry/src/main/java/io/sentry/SentryClient.java#L540-L543

I'm still not sure what the original request of this issue is: would you like to distinguish between user-perceived crashes vs not, or would you like us to fix the coroutine exception handling? I'm a bit lost I guess, sorry

**Comment by arifken** [comment]
The primary goal is to be able to distinguish between user-perceived crashes vs. not (ideally some sort of a query option we could use in discover, dashboards, and alerts)

**Comment by anthonycr** [comment]
> I'm not sure if this is correct, because we actually finish the session if our exception handler encountered a crash, see here

we were seeing a scenario where the crash was actually happening while the app was in the foreground, but the application itself was not terminated. that is, Activity 2 was stacked on top of Activity 1, and a user interaction caused a crash in Activity 2, which triggered Activity 1 to restart. For some reason, it seems that is not treated as a "user perceived" crash. 

Maybe there's a way for us to create a sample project to demonstrate the issue, would that be helpful? The google docs on user perceived crashes seem clear, but I think they're obfuscating what the Android OS is actually doing.

> You can query for fatal crashes happening while the app was in foreground (which is the main factor used to define user perceived) like so: level:fatal app.in_foreground:True. Does this help?

I'll defer to @arifken on this.

**Comment by romtsn** [comment]
> Maybe there's a way for us to create a sample project to demonstrate the issue, would that be helpful? The google docs on user perceived crashes seem clear, but I think they're obfuscating what the Android OS is actually doing.

Yeah, a sample project is always welcome, would definitely help us investigate, thank you!

**Comment by markushi** [comment]
We definitely can look into the new-ish behavior when multiple activities are on the stack and a crash just closes the topmost one, or in other words the app is not terminated.

**Comment by anthonycr** [comment]
Attached is a zip file of the test case. The example I created is an app consisting of 2 activities.
- MainActivity
- SecondaryActivity

`MainActivity` has 2 buttons, a `Crash` button, and a `Second Screen` button. The `SecondaryActivity` has a `Crash` button. Both `Crash` buttons do the same thing, invoke the kotlin function `error("oops")`, which will throw an exception on the current thread. The `Second Screen` button will open `SecondaryActivity`.

When the `Crash` button is clicked on `MainActivity`, the app will close and show that it's "crashed." The app process is terminated, and is not restarted (Check logcat for `PROCESS ENDED` and `PROCESS STARTED`). This presumably ends the session and is counted as a user visible crash.

When the `Crash` button is clicked on `SecondaryActivity`, the app will not close. Instead, the screen goes blank for a second and does not show the app crash dialog. Instead, what happens is that the `MainActivity` is shown again. Looking in logcat, we can see that the app process is terminated, and immediately recreated without any user intervention. I think this crash is being reported as a background crash. What's interesting is that the behavior is different if I trigger crash 1 before crash 2. If I trigger crash 1 first, then crash 2 seems to terminate the app without restarting it.

[Crash.zip](https://github.com/user-attachments/files/15957822/Crash.zip)

**Comment by romtsn** [comment]
@anthonycr @arifken sorry for abandoning this, but I'm reviving this now.

> Looking in logcat, we can see that the app process is terminated, and immediately recreated without any user intervention. I think this crash is being reported as a background crash. 

But this is even more weird to me. If the process gets terminated and recreated, how come this can be a background crash for google play? It makes little sense to me. Would you folks consider this crash user-perceived or actually background in this case?
